### PR TITLE
KEYCLOAK-18552 - Grouping of attributes in user profile (step 1)

### PR DESCRIPTION
**Pull request by ioemat**
Group attributes in user profile dialog, show group header and an optional description.

Attributes are grouped by adding the annotation _group_ with the group name as value.

Group header and optional description are configured via realm localization using the keys
`userprofile.attributegroup.header.<groupname>` and `userprofile.attributegroup.description.<groupname>`.

See also [KEYCLOAK-18552](https://issues.redhat.com/browse/KEYCLOAK-18552).

### CODE REVIEW

**Comment by velias** [comment]
Thanks @ioemat . It would be good to wait until PR #8222 is merged, as it refactors templates for pages containing dynamic user profile form a bit. Dynamic rendering is extracted to separate ftl and shared by all page templates this way. So your change will be simpler then.

**Comment by velias** [comment]
One thing is how to approach i18n. For attributes, admin can use `displayName` to either directly enter text used in template (great for simple cases where i18n not necessary), or can use placeholders to lookup texts from i18n resources. We should probably stay consistent in this are and use the same mechanism for group header and description.

**Comment by ioemat** [comment]
Thanks for the feedback, I agree that it makes sense to rebase my changes on the mentioned PR.

To make things more clear a took a screenshot of what this PR adds (note the "Address" header and description below) 

2021-07-02_11h56_24


Regarding i18n things are probably a little bit more complicated since the intention of the group annotation is to be used on multiple attributes, e.g.

```
{
  "attributes" : [
    {
        "name": "street",
        "annotations": {
            "group": "address"
        }
    },
    {
        "name": "city",
        "annotations": {
            "group": "address"
        }
    }
  ...
  ]
}
```

When introducing an additional `displayName` this would lead to a configuration like this:

```
{
  "attributes" : [
    {
        "name": "street",
        "annotations": {
            "group": "address",
            "groupDisplayName" : "Address"
        }
    },
    {
        "name": "city",
        "annotations": {
            "group": "address",
            "groupDisplayName" : "Address"
        }
    },
  ...
  ]
}
```

Which may be a confusing to the user, especially when the text assigned to `groupDisplayName` is different for the same group.
That's why I decided to follow the dynamic approach that is implemented in the PR.

To avoid having two annotations that are related to each other I would rather make `group` an attribute meta data e.g. like this

```
{
  "attributes" : [
    {
        "name": "street",
        "group: {
            "name": "address",
            "header" : "Address",
            "description" : "Your address is required for ...",
        }
    },
    {
        "name": "city",
        "group": {
            "name": "address"
        }
    },
  ...
  ]
}
```
`header`  and `description` should be optional parameters that only have an effect when set for the first group member (in this example `street`). Here one can also use i18n placeholders like this:


```
{
  "attributes" : [
    {
        "name": "street",
        "group: {
            "name": "address",
            "header" : "${address.header}",
            "description" : "${address.description}",
        }
    },
    {
        "name": "city",
        "group": {
            "name": "address"
        }
    },
  ...
  ]
}
```

What do you think about this @velias?

**Comment by velias** [comment]
> header and description should be optional parameters that only have an effect when set for the first group member
This may be a problem when some fields disappear due to `selector` or `permissions`. 
So if we would introduce special handling of groups, it would be probably better to have extra area to define groups with their properties, and then only reference them from attributes. But it makes things really complicated. So maybe your current impl with explicit i18n keys will be the simplest so best ;-)

**Comment by pedroigor** [comment]
@ioemat @velias Don't you guys feel that a top-level group property makes sense here?

You would just define the group once and then reference the group through a group property.

**Comment by ioemat** [comment]
@pedroigor you're thinking about an `groups` element in parallel to `attributes` like this?

```
{
 "groups" : [
        "group: {
            "name": "address",
            "header" : "${address.header}",
            "description" : "${address.description}",
        }
  ],
  "attributes" : [
    {
        "name": "street",
        "group: "address",
        "displayName" : "${address.street}"
        ...
    {
        "name": "city",
        "group":  "address",
        "displayName" : "${address.city}"
        ...
    },
  ...
  ]
}
```

This would be of course a very clean solution to solve the issues mentioned above.

**Comment by pedroigor** [comment]
@ioemat Yeah. I would just remove the top-level `group` property and have group properties similar to how we define attributes to keep consistency.

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributeGroups.java
Comment (ioemat) [inline server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributeGroups.java:41]: We had a discussion internally whether this class should be better named `AttributeGroupsImpl` - if so the rename should probably also been applied to `DefaultAttributes`. @pedroigor  @velias  what do you think?

**Comment by ioemat** [comment]
@pedroigor  @velias pushed an update implementing the changes we discussed, looking forward to your feedback :)

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributeGroups.java
Comment (pedroigor) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributeGroups.java:41]: I'm not sure if we need a specific hierarchy/abstraction for groups. We should be fine by just leveraging the `Attributes` interface. 

It should make things more simple.

**Comment by pedroigor** [comment]
@ioemat What about the admin console UI changes? Do we want to enable now managing groups only through the JSON editor?

**Comment by ioemat** [comment]
@pedroigor I didn't want so spent much effort for the UI part before we haven't agreed on the backend functionality.
I can contribute the admin console ui part as well, but suggest to have an own PR for this, are you fine with this?

**Comment by pedroigor** [comment]
@ioemat I'm OK doing as a follow-up and rely on JSON editor for group configuration.

**Comment by velias** [comment]
@ioemat please implement validation of groups config after reading from json (look at attribute config validation in config parser). Validation have to make sure that group has name (and maybe displayName, depends how you handle it in template), but mainly make sure attribute references existing group. Thanks

**Comment by ioemat** [comment]
@velias thanks for the hint, I added the validation and tests for it.

File: services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java:86]: Do we really need two separate validation methods? Looks they are called from one place, so I'd say we can have only one `validate()` as before doing all necessary validations.

File: server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java
Comment (velias) [inline server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java:27]: Do we really need all this completely separate and complex handling of groups? For me they are only additional info for GUI rendering. Much easier/simpler would be to keep complete group info in `AttributeMetdata` of individual attribute, and let `org.keycloak.userprofile.DeclarativeUserProfileProvider.decorateUserProfileForCache(UserProfileMetadata, ComponentModel)` to set it into each AttributeMetadata based on parsed configuration.

File: services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java
Comment (ioemat) [reply services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java:86]: yep, makes sense. I'll create a validate() method with calls the two methods and aggregates the result.

File: server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java
Comment (ioemat) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java:27]: ok, since @pedroigor mention something similar above I'm going to adapt the code, will push an updated version probably on monday

File: services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java
Comment (velias) [reply services/src/main/java/org/keycloak/userprofile/config/UPConfigUtils.java:86]: thanks

File: server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java
Comment (velias) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java:27]: thanks, I think it will be much simpler code

Comment (velias) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java:27]: thanks

Comment (ioemat) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AttributeGroupMetadata.java:28]: @velias  @pedroigor, I pushed an update that provides group metadata as part of the attribute metadata

File: services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java
Comment (velias) [inline services/src/main/java/org/keycloak/userprofile/DeclarativeUserProfileProvider.java:268]: I do not see any need for new method here, simply take groups from the config and use them in attributes config loop inside of the `decorateUserProfileForCache()` method

File: server-spi-private/src/main/java/org/keycloak/userprofile/AttributeMetadata.java
Comment (velias) [inline server-spi-private/src/main/java/org/keycloak/userprofile/AttributeMetadata.java:46]: group name is inside of the `AttributeGroupMetadata`, so separate field `attributeGroup` is not necessary here I think

**Comment by velias** [comment]
Thanks @ioemat, implementation looks very good to me now, only two small remarks added. 

I also just realized that there is not any test checking that group are really rendered into the html, it would be good to have this kind of test also. Small problem is that dynamic user profile form is used for more pages, so this test should be added for all of them (but it will be very similar for all of them). There are relevant test classes for all relevant pages:
* `/integration-arquillian-tests-base/src/test/java/org/keycloak/testsuite/forms/RegisterWithUserProfileTest.java`
* `/integration-arquillian-tests-base/src/test/java/org/keycloak/testsuite/forms/VerifyProfileTest.java`
* `/integration-arquillian-tests-base/src/test/java/org/keycloak/testsuite/actions/RequiredActionUpdateProfileWithUserProfileTest.java`
* `/integration-arquillian-tests-base/src/test/java/org/keycloak/testsuite/broker/KcOidcFirstBrokerLoginWithUserProfileTest.java`

Also testing itself can be a bit problematic, you probably have to use CSS selectors to check/assert content of the HTML. For inspiration see method `testAttributeGuiOrder()` present in all of these tests. You can see how to set configuration, open given page and how to assert its content using CSS selectors here.

**Comment by ioemat** [comment]
@velias - I fixed the findings you mentioned above and added the UI tests. Thanks for pointing me to all the test classes.

**Comment by velias** [comment]
@ioemat thanks, but looks like some of your new tests are failing `KcOidcFirstBrokerLoginWithUserProfileTest.testAttributeGrouping:101 » NoSuchElement`

**Comment by ioemat** [comment]
@velias I rebase the code and fixed the test, please have a look again

**Comment by velias** [comment]
@ioemat thx, LGTM, @pedroigor

**Review by velias — APPROVED** [review]

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@ioemat @velias Thanks.

@ioemat Created https://issues.redhat.com/browse/KEYCLOAK-18875.

**Review by stianst — APPROVED** [review]

**Comment by pedroigor** [comment]
@ioemat Created https://issues.redhat.com/browse/KEYCLOAK-18875 as a follow-up.

**Pull request merged**
